agenda/data/src/main/java/com/nibalk/tasky/agenda/data/OfflineFirstAgendaRepository.kt

```diff
@@ -79,4 +76,31 @@ class OfflineFirstAgendaRepository(
             }
         }
     }
+
+    override suspend fun fetchAllAgendas(): EmptyResult<DataError> {
+        return when (val remoteResult = remoteAgendaDataSource.getFullAgenda()) {
+            is Result.Success -> {
+                val agendaItems  = remoteResult.data
+                if (agendaItems != null) {
+                    applicationScope.async {
+                        localEventDataSource.upsertEvents(agendaItems.events)
+                            .onError { localError -> Result.Error(localError) }
+                            .asEmptyDataResult()
+                        localTaskDataSource.upsertTasks(agendaItems.tasks)
+                            .onError { localError -> Result.Error(localError) }
+                            .asEmptyDataResult()
+                        localReminderDataSource.upsertReminders(agendaItems.reminders)
+                            .onError { localError -> Result.Error(localError) }
+                            .asEmptyDataResult()
```

Comment:
Here, I would recommend to use a database transaction instead. That will ensure that all these DB calls will only be applied to the DB if they all succeed. If one of these wouldn't succeed for whatever reason (out of storage, coroutine gets cancelled, etc.), your DB will be in an inconsistent state. The first calls may have already succeeded, but the later calls will never be executed due to a cancellation for example.
In Room, you can create a transaction (just as an example, needs to be adjusted to your use case):
```
@Dao
interface AgendaDao {

    @Transaction
    suspend fun updateAgenda(db: AgendaDatabase, agendaItems: List<AgendaItem>) {
        db.apply {
            eventDao.deleteAllEvents()
            attendeeDao.deleteAllAttendees()
            taskDao.deleteAllTasks()
            reminderDao.deleteAllReminders()

            val events = agendaItems.filterIsInstance<AgendaItem.Event>()
            val tasks = agendaItems.filterIsInstance<AgendaItem.Task>()
            val reminders = agendaItems.filterIsInstance<AgendaItem.Reminder>()
            events.forEach {
                eventDao.insertEvent(it.toEventEntity())
                it.attendees.forEach {
                    attendeeDao.insertAttendee(it.toAttendeeEntity)
                }
            }
            tasks.forEach {
                taskDao.insertTask(it.toTaskEntity())
            }
            reminders.forEach {
                reminderDao.insertReminder(it.toReminderEntity())
            }
        }
    }
}
```
Whenever you need to execute multiple DB calls that should only be applied to the DB if they all succeed, you need something like this.